File: src/oop/delegation2/delegation.java
class A {
    public void foo() {
        System.out.print("a.foo ");
        bar();
    }

    public void bar() {
        System.out.println("a.bar");
    }
}

class B extends A {
    public void foo() {
        System.out.print("b.foo ");
        super.foo();
    }

    public void bar() {
        System.out.println("b.bar");
    }
}

public class Main {
    public static void main(String[] args) {
        A a = new A();
        a.foo();

        B b = new B();
        b.foo(); // "b.foo a.foo a.bar" or "b.foo a.foo b.bar"?
    }
}

File: src/oop/delegation2/delegation.go
package main

import "fmt"

type A struct{}

func (a A) Foo() {
	fmt.Print("a.foo ")
	a.Bar()
}

func (a A) Bar() {
	fmt.Println("a.bar")
}

type B struct {
	A
}

func (b B) Foo() {
	fmt.Print("b.foo ")
	b.A.Foo()
}

func (b B) Bar() {
	fmt.Println("b.bar")
}

func main() {
	a := A{}
	a.Foo()

	b := B{}
	b.Foo() // "b.foo a.foo a.bar" or "b.foo a.foo b.bar"?
}
